File: crates/jellyfish-merkle/src/proof.rs
use {
    crate::{hash_internal_node, hash_leaf_node, hash_of, BitArray, InternalNode, LeafNode, Node},
    cw_std::{cw_serde, Hash},
    thiserror::Error,
};

#[derive(Debug, Error)]
pub enum ProofError {
    #[error("incorrect proof type, expect {expect}, got {actual}")]
    IncorrectProofType {
        expect: &'static str,
        actual: &'static str,
    },

    #[error("root hash mismatch! computed: {computed}, actual: {actual}")]
    RootHashMismatch {
        computed: Hash,
        actual:   Hash,
    },

    // TODO: add more details to the error message?
    #[error("expecting child to not exist but it exists")]
    UnexpectedChild,

    // TODO: add more details to the error message?
    #[error("expecting bitarrays to share a common prefix but they do not")]
    NotCommonPrefix,
}

#[cw_serde]
pub enum Proof {
    Membership {
        sibling_hashes: Vec<Option<Hash>>,
    },
    NonMembership {
        node: Node,
        sibling_hashes: Vec<Option<Hash>>,
    },
}

pub fn verify_membership(
    root_hash:  &Hash,
    key_hash:   &Hash,
    value_hash: &Hash,
    proof:      &Proof,
) -> Result<(), ProofError> {
    let Proof::Membership { sibling_hashes } = proof else {
        return Err(ProofError::IncorrectProofType {
            expect: "membership",
            actual: "non-membership",
        });
    };

    let bitarray = BitArray::from_bytes(key_hash);
    let hash = hash_leaf_node(key_hash, value_hash);

    compute_and_compare_root_hash(root_hash, &bitarray, sibling_hashes, hash)
}

pub fn verify_non_membership(
    root_hash: &Hash,
    key_hash:  &Hash,
    proof:     &Proof,
) -> Result<(), ProofError> {
    let Proof::NonMembership { node, sibling_hashes } = proof else {
        return Err(ProofError::IncorrectProofType {
            expect: "non-membership",
            actual: "membership",
        });
    };

    let bitarray = BitArray::from_bytes(key_hash);
    let hash = match node {
        // if the node given is an internal node, we check the bit at the depth.
        // if the bit is a 0, it must not have a left child; if the bit is a 1,
        // it must not have a right child.
        Node::Internal(InternalNode { left_child, right_child }) => {
            match (bitarray.bit_at_index(sibling_hashes.len()), left_child, right_child) {
                (0, Some(_), _) | (1, _, Some(_)) => {
                    return Err(ProofError::UnexpectedChild);
                },
                _ => hash_internal_node(hash_of(left_child), hash_of(right_child)),
            }
        },
        // if the node given is a leaf, it's bit path must share a common prefix
        // with the key we want to prove not exist.
        Node::Leaf(LeafNode { key_hash, value_hash }) => {
            let non_exist_bitarray = BitArray::from_bytes(key_hash);
            let exist_bits = bitarray.reverse_iterate_from_index(sibling_hashes.len());
            let non_exist_bits = non_exist_bitarray.reverse_iterate_from_index(sibling_hashes.len());
            if exist_bits.zip(non_exist_bits).any(|(a, b)| a != b) {
                return Err(ProofError::NotCommonPrefix);
            }
            hash_leaf_node(key_hash, value_hash)
        },
    };

    compute_and_compare_root_hash(root_hash, &bitarray, sibling_hashes, hash)
}

fn compute_and_compare_root_hash(
    root_hash: &Hash,
    bitarray: &BitArray,
    sibling_hashes: &[Option<Hash>],
    mut hash: Hash,
) -> Result<(), ProofError> {
    for (bit, sibling_hash) in bitarray.reverse_iterate_from_index(sibling_hashes.len()).zip(sibling_hashes) {
        if bit == 0 {
            hash = hash_internal_node(Some(&hash), sibling_hash.as_ref());
        } else {
            hash = hash_internal_node(sibling_hash.as_ref(), Some(&hash));
        }
    }

    if hash != *root_hash {
        return Err(ProofError::RootHashMismatch {
            computed: hash,
            actual: root_hash.clone(),
        });
    }

    Ok(())
}

// ----------------------------------- tests -----------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn proving_membership() {
        todo!()
    }

    #[test]
    fn proving_non_membership() {
        todo!()
    }
}

File: sdk/packages/core/src/actions/public/proof.ts
import { Sha256 } from "@left-curve/crypto";
import { decodeHex, encodeHex } from "@left-curve/encoding";
import type { MembershipProof, NonMembershipProof, Proof } from "@left-curve/types";

export function verifyProof(
  rootHash: Uint8Array,
  keyHash: Uint8Array,
  valueHash: Uint8Array | null,
  proof: Proof,
) {
  // value exists, the proof must be a membership proof
  if (valueHash !== null) {
    if ("membership" in proof) {
      return verifyMembershipProof(rootHash, keyHash, valueHash, proof.membership);
    }
    throw new Error("expecting membership proof, got non-membership proof");
  }
  // value doesn't exist, the proof must be a non-membership proof
  if ("nonMembership" in proof) {
    return verifyNonMembershipProof(rootHash, keyHash, proof.nonMembership);
  }
  throw new Error("expecting non-membership proof, got membership proof");
}

export function verifyMembershipProof(
  rootHash: Uint8Array,
  keyHash: Uint8Array,
  valueHash: Uint8Array,
  proof: MembershipProof,
) {
  const hash = hashLeafNode(keyHash, valueHash);
  return computeAndCompareRootHash(rootHash, keyHash, proof.siblingHashes, hash);
}

export function verifyNonMembershipProof(
  rootHash: Uint8Array,
  keyHash: Uint8Array,
  proof: NonMembershipProof,
) {
  let hash: Uint8Array;
  if ("internal" in proof.node) {
    const { leftHash, rightHash } = proof.node.internal;
    const bit = getBitAtIndex(keyHash, proof.siblingHashes.length);
    if ((bit === 0 && !!leftHash) || (bit === 1 && !!rightHash)) {
      throw new Error("expecting child to not exist but it exists");
    }
    hash = hashInternalNode(decodeNullableHex(leftHash), decodeNullableHex(rightHash));
  } else {
    const existingKeyHash = decodeHex(proof.node.leaf.keyHash);
    for (let i = proof.siblingHashes.length - 1; i >= 0; i--) {
      if (getBitAtIndex(keyHash, i) !== getBitAtIndex(existingKeyHash, i)) {
        throw new Error("expecting bitarrays to share a common prefix but they do not");
      }
    }
    hash = hashLeafNode(existingKeyHash, decodeHex(proof.node.leaf.valueHash));
  }
  return computeAndCompareRootHash(rootHash, keyHash, proof.siblingHashes, hash);
}

function computeAndCompareRootHash(
  rootHash: Uint8Array,
  keyHash: Uint8Array,
  siblingHashes: (string | null)[],
  hash: Uint8Array,
) {
  for (let i = 0; i < siblingHashes.length; i++) {
    if (getBitAtIndex(keyHash, siblingHashes.length - i - 1) === 0) {
      hash = hashInternalNode(hash, decodeNullableHex(siblingHashes[i]));
    } else {
      hash = hashInternalNode(decodeNullableHex(siblingHashes[i]), hash);
    }
  }

  for (let i = 0; i < 32; i++) {
    if (rootHash[i] !== hash[i]) {
      throw new Error(
        `root hash mismatch! computed: ${encodeHex(hash)}, actual: ${encodeHex(rootHash)}`,
      );
    }
  }
}

function hashInternalNode(leftHash: Uint8Array | null, rightHash: Uint8Array | null): Uint8Array {
  const hasher = new Sha256();
  hasher.update(new Uint8Array([0])); // internal node prefix
  if (leftHash !== null) {
    hasher.update(leftHash);
  } else {
    hasher.update(new Uint8Array(32)); // this creates an all-zero byte array
  }
  if (rightHash !== null) {
    hasher.update(rightHash);
  } else {
    hasher.update(new Uint8Array(32));
  }
  return hasher.digest();
}

function hashLeafNode(keyHash: Uint8Array, valueHash: Uint8Array): Uint8Array {
  const hasher = new Sha256();
  hasher.update(new Uint8Array([1])); // leaf node prefix
  hasher.update(keyHash);
  hasher.update(valueHash);
  return hasher.digest();
}

function getBitAtIndex(hash: Uint8Array, index: number): number {
  const quotient = Math.floor(index / 8);
  const remainder = index % 8;
  const byte = hash[quotient];
  return (byte >> (7 - remainder)) & 1;
}

function decodeNullableHex(hexStr: string | null): Uint8Array | null {
  return hexStr !== null ? decodeHex(hexStr) : null;
}
